src/rgw/rgw_rest.cc

```diff
@@ -1674,6 +1692,32 @@ int RGWREST::preprocess(struct req_state *s, RGWClientIO* cio)
       }
     }
 
+    // Handle A/CNAME records that point to the RGW storage, but do match the
+    // CNAME test above, per issue http://tracker.ceph.com/issues/15975
+    // If BOTH domain & subdomain variables are empty, then none of the above
+    // cases matched anything, and we should fall back to using the Host header
+    // directly as the bucket name.
+    // As additional checks:
+    // - if the Host header is an IP, we're using path-style access without DNS
+    // - Also check that the Host header is a valid bucket name (per relaxed
+    //   bucket rules) before using it.
+    if (subdomain.empty()
+        && (domain.empty() || domain != info.host)
+        && !looks_like_ip_address(info.host.c_str())
+        && valid_s3_bucket_name(info.host, true)) {
```

Comment:
@oritwas you mean RGWHandler_REST::validate_bucket_name validate_bucket_name which isn't static; I'll tweak it so it can be used in the static context of RGWREST::preprocess.